controller_manager/test/test_controller_manager_srvs.cpp

```diff
@@ -480,3 +480,574 @@ TEST_F(TestControllerManagerSrvs, configure_controller_srv)
+  chained_cmd_cfg = {
+    controller_interface::interface_configuration_type::INDIVIDUAL,
+    {std::string(TEST_CHAINED_CONTROLLER_6) + "/joint1/position",
+     std::string(TEST_CHAINED_CONTROLLER_6) + "/joint2/velocity"}};
+  test_chained_controller_7->set_command_interface_configuration(chained_cmd_cfg);
+  test_chained_controller_7->set_state_interface_configuration(chained_state_cfg);
+  test_chained_controller_7->set_reference_interface_names({"joint1/position", "joint2/velocity"});
+
+  // create non-chained controller
+  auto test_controller = std::make_shared<TestController>();
+  controller_interface::InterfaceConfiguration cmd_cfg = {
+    controller_interface::interface_configuration_type::INDIVIDUAL,
+    {std::string(TEST_CHAINED_CONTROLLER_7) + "/joint1/position",
+     std::string(TEST_CHAINED_CONTROLLER_7) + "/joint2/velocity"}};
+  controller_interface::InterfaceConfiguration state_cfg = {
+    controller_interface::interface_configuration_type::INDIVIDUAL,
+    {"joint1/position", "joint2/velocity"}};
+  test_controller->set_command_interface_configuration(cmd_cfg);
+  test_controller->set_state_interface_configuration(state_cfg);
+  // add controllers
+  cm_->add_controller(
+    test_chained_controller_4, TEST_CHAINED_CONTROLLER_4,
+    test_chainable_controller::TEST_CONTROLLER_CLASS_NAME);
+  cm_->add_controller(
+    test_chained_controller_1, TEST_CHAINED_CONTROLLER_1,
+    test_chainable_controller::TEST_CONTROLLER_CLASS_NAME);
+  cm_->add_controller(
+    test_chained_controller_3, TEST_CHAINED_CONTROLLER_3,
+    test_chainable_controller::TEST_CONTROLLER_CLASS_NAME);
+  cm_->add_controller(
+    test_chained_controller_5, TEST_CHAINED_CONTROLLER_5,
+    test_chainable_controller::TEST_CONTROLLER_CLASS_NAME);
+  cm_->add_controller(
+    test_chained_controller_6, TEST_CHAINED_CONTROLLER_6,
+    test_chainable_controller::TEST_CONTROLLER_CLASS_NAME);
+  cm_->add_controller(
+    test_chained_controller_2, TEST_CHAINED_CONTROLLER_2,
+    test_chainable_controller::TEST_CONTROLLER_CLASS_NAME);
+  cm_->add_controller(
+    test_controller, test_controller::TEST_CONTROLLER_NAME,
+    test_controller::TEST_CONTROLLER_CLASS_NAME);
+  cm_->add_controller(
+    test_chained_controller_7, TEST_CHAINED_CONTROLLER_7,
+    test_chainable_controller::TEST_CONTROLLER_CLASS_NAME);
+  // get controller list before configure
+  auto result = call_service_and_wait(*client, request, srv_executor);
+  // check chainable controller
+  ASSERT_EQ(8u, result->controller.size());
+  ASSERT_EQ(result->controller[0].name, TEST_CHAINED_CONTROLLER_4);
+  ASSERT_EQ(result->controller[1].name, TEST_CHAINED_CONTROLLER_1);
+  ASSERT_EQ(result->controller[2].name, TEST_CHAINED_CONTROLLER_3);
+  ASSERT_EQ(result->controller[3].name, TEST_CHAINED_CONTROLLER_5);
+  ASSERT_EQ(result->controller[4].name, TEST_CHAINED_CONTROLLER_6);
+  ASSERT_EQ(result->controller[5].name, TEST_CHAINED_CONTROLLER_2);
```

Comment:
I am bit confused here- from the description I would expect "test_controller_name" and "controller_6" to be at the begining.

Reply:
It is related to what I have explained in the other comment, as I am unable to call the `command_interface_onfiguration` and `state_interface_configuration` in any other state other than `ACTIVE` or `INACTIVE` state,

The sorting only happens when the controllers are once activated. This check you are referring to is right after adding to the controller list and also when they are preconfigured. This exactly represents the order in which the controllers have added above with `add_controller`, the check of the sorted controller is here:

https://github.com/ros-controls/ros2_control/blob/3ef00f3386d8b194f133f8936f764e15c218391a/controller_manager/test/test_controller_manager_srvs.cpp#L787-L818